Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Predicting the number of hours worked: a neural network vs logistic model approach"
COLLABORATORS = "Matteo Turrini and Manuel Espinosa"

---

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Datascience-for-Economics-Final-Assignment" data-toc-modified-id="Datascience-for-Economics-Final-Assignment-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Datascience for Economics Final Assignment</a></span></li><li><span><a href="#Research-question-(0.5-points)" data-toc-modified-id="Research-question-(0.5-points)-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Research question (0.5 points)</a></span></li><li><span><a href="#Motivation-(0.5-points)" data-toc-modified-id="Motivation-(0.5-points)-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Motivation (0.5 points)</a></span></li><li><span><a href="#Method-and-data-(1.0-point)" data-toc-modified-id="Method-and-data-(1.0-point)-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Method and data (1.0 point)</a></span></li><li><span><a href="#Preview-of-the-answers-(0.5-points)" data-toc-modified-id="Preview-of-the-answers-(0.5-points)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Preview of the answers (0.5 points)</a></span></li><li><span><a href="#Main-assumptions-(0.5-points)" data-toc-modified-id="Main-assumptions-(0.5-points)-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Main assumptions (0.5 points)</a></span></li><li><span><a href="#Python/R-code-(6-points)" data-toc-modified-id="Python/R-code-(6-points)-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Python/R code (6 points)</a></span></li><li><span><a href="#Robustness-analysis-(0.5-points)" data-toc-modified-id="Robustness-analysis-(0.5-points)-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Robustness analysis (0.5 points)</a></span></li><li><span><a href="#Discussion-and-conclusion-(0.5-points)" data-toc-modified-id="Discussion-and-conclusion-(0.5-points)-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Discussion and conclusion (0.5 points)</a></span></li></ul></div>

For this problem set, we'll be using the Jupyter notebook:

![](jupyter.png)

# Datascience for Economics Final Assignment



|Name|SNR|ANR|
|----|---|----|
|Matteo Turrini|2084906|u207794|
|Manuel Espinosa|2067814|u112479|




# Research question (0.5 points)



How does personal characteristics help predicting the level of hours worked per week in the UK? Which method performs better in doing so: Neural networks or Logistic Multinomial Regressions?

# Motivation (0.5 points)



Predicting the number of hours worked per week could be a helpful tool in order to classify workers on the basis of this parameter, and then use this information to promote and implement Active Labor Market Programmes (ALMPs). These are government programmes aimed at increasing the rate of employment by bringing unemployed people into employment, preventing already employed people to lose their jobs, and increasing productivity and earnings: this way, governments are able to improve the functioning of labor markets. Therefore, predicting the hours worked per week per type of worker can provide some "profiles" of workers that ALMPs can use to improve the efficiency of a working sector or to maintain an already efficient one.

**Moreover, we think that it would be interesting to compare two methods widely used in classification problems like ours: on one hand, we have  Neural networks, which are capable of disentangling different patterns in the data by using non-linear functions like relu and sigmoid [(Dreiseitl  et al., 2012)](https://doi.org/10.1016/S1532-0464(03)00034-0); on the other hand, we have Logistic Multinomial Regression, which is another classification method used to analyse the relationship between the outcome and multiple independent variables. **

# Method and data (1.0 point)



The data we have used to conduct the study comes from the English Office for National Statistics, here's the link to the [website](https://www.ons.gov.uk/peoplepopulationandcommunity/educationandchildcare/datasets/2011censusteachingfile). 
    This is a sample of 2011 providing microdata with 569741 observations. People analysed in this sample are classified under 16 cateorical variables:the region, the residence type, family composition, population base, sex, age, marital status, student, country of birth, health, ethnic group, religion, economic activity, occupation, industry, hours worked per week, and approximated social grade. For a more in depth explanation of each variable and the different values it can take you can access this [file](Microdatateachingvariablels.pdf). This dataset is going to be [cleaned](#data_cleaning) and [transformed](#data_transformation) using pandas library. Once the data is ready there will be a brief descriptive analysis of our variables to see any relevant distribution or relationship.
    
   Regarding the methodology of our study it will consist in predicting, following a classification problem, the hours worked per week of our individuals. In order to do so, we first have to prepare our features and our target variable and have them separate. Having done this, we will first apply a [Neural network approach](#neural_network), in which we will compare several neural networks by implementing a simple parameter tunning technique; Then, we will implement a [logistic multinomial regression](#logistic_regression) comparing a few of them by introducing the penalty term. Finally, we will choose the best model for each approach based on its performance and compare them by implementing a [bootstrapping](#bootstrapping) tool that will serve as a robustness analysis. 

# Preview of the answers (0.5 points)



Since the research question is divided in two we considerthat the preview of the answer should be divided as well. There for the main results of our study have been:
* Our best model has a prediction accuracy of the level of hours worked per week that goes from 78% to 78,5% in the 95% of times.
* The Neural Network approach that a slightly better job than the Logistic Nominal Regression in predicting the level of hours worked per week. This slightly difference can be quantified as a 0.5% in the prediction accuracy.

However, we suggest you to read our discussion to understand that our results are not robust enough.

# Main assumptions (0.5 points)


In order to answer to the research question with our analysis, we need to make two main assumptions: first of all, we assume the survey to be randomized in order to minimize data corruption, avoiding bias, and improve data collection. Our other main assumption is that we assume, in order for this model to be valid, that there are no significant changes in the economic environment, the industry or the markets: for this reason, we can't guarantee that this model will hold in the future if there will be significant changes in the aforementioned fields. 
Given that we don't establish any causal relationship in the model, there's no need to make additional assumptions regarding the distribution of variables and we do not have to worry about facing any problem such as missing counfounders, multicolinearity between variables or omitted variable bias. 


# Python/R code (6 points)



## Package importation

First of all we need to import the libraries for the functions we are going to use.

In [2]:
#Imports

#we import the numpy library  for many uses such as generate new data or manipulate arrays
import numpy as np
#the pandas library to manipulate and transform the data
import pandas as pd
#the seaborn library which is based on matplotlib and makes the graphs more attractive
import seaborn as sn



#matplotlib and concretely pyplot will be useful to plot the different graphs that we need to show our results
import matplotlib.pyplot as plt
#is a library designed for machine learning and it contains keras which we will use for the neural network
import tensorflow as tf
from tensorflow.keras import models
from tensorflow.keras import layers

#we will need the LogisticRegression model to implement the second approach
from sklearn.linear_model import LogisticRegression





#we will use it to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')
#we import scikit learn to access different statistical tools and functions such as methods to split the data in training and testing set, 
#to standardize the data if needed and to compute the accuracy between predicted and true values
#from sklearn.preprocessing import scale
#from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


#We finally import the ipywidgets package to make our functions interactive and display them
from ipywidgets import interact, interactive, fixed, Layout
import ipywidgets as widgets 
from IPython.display import display

## Data manipulation

We will have to check thoroughly the structure of the data as well as the column types to see how to transform the data in a way that can be used afterwards, taking into account that all of the variables are categorical.

### Data importation

We first import the data from a csv that you will find in the [repository](/).

In [3]:
#Data importation
#set the file directory path
file=('2011 Census Microdata Teaching File.csv')
#use the pandas methos read_csv to create a dataframe from the csv
df= pd.read_csv(file, header=None)
#we set the column names
df.columns = df.iloc[1,:].values
#we drop the first two raws of the file as there is no interesting information in these.
df.drop([0,1], axis=0, inplace=True)
#after manipulating the dataframe, we reset the index of it
df_small=df.reset_index(drop=True)
df_small


Person ID     Region Residence Type Family Composition Population Base  \
0        7394816  E12000001              H                  2               1   
1        7394745  E12000001              H                  5               1   
2        7395066  E12000001              H                  3               1   
3        7395329  E12000001              H                  3               1   
4        7394712  E12000001              H                  3               1   
...          ...        ...            ...                ...             ...   
569736   7946020  W92000004              H                  1               1   
569737   7944310  W92000004              H                  3               1   
569738   7945374  W92000004              H                  3               1   
569739   7944768  W92000004              H                  1               1   
569740   7944959  W92000004              H                  2               1   

       Sex Age Marital Status Student Country of Birth Health Ethnic Group  \
0        2   6              2       2                1      2            1   
1        1   4              1       2                1      1            1   
2        2   4              1       2                1      1            1   
3        2   2              1       2                1      2            1   
4        1   5              4       2                1      1            1   
...     ..  ..            ...     ...              ...    ...          ...   
569736   1   5              1       2                1      4            1   
569737   1   3              1       2                1      2            1   
569738   1   1              1       1                1      1            1   
569739   2   8              5       2                1      3            1   
569740   2   2              2       2                1      2            1   

       Religion Economic Activity Occupation Industry Hours worked per week  \
0             2                 5          8        2                    -9   
1             2                 1          8        6                     4   
2             1                 1          6       11                     3   
3             2                 1          7        7                     3   
4             2                 1          1        4                     3   
...         ...               ...        ...      ...                   ...   
569736        9                 1          8        8                     3   
569737        1                 1          7        4                     3   
569738        2                -9         -9       -9                    -9   
569739        9                 5          9        2                    -9   
569740        1                 1          7        4                     1   

       Approximated Social Grade  
0                              4  
1                              3  
2                              4  
3                              2  
4                              2  
...                          ...  
569736                         3  
569737                         4  
569738                        -9  
569739                         4  
569740                         4  

[569741 rows x 18 columns]

### Data transformation  <a id='data_transformation'></a>

Given that all of the variables are integer encoded except from region and Residence type, we preferred to change these two to integer encoded as well. Moreover, all the columns were of the type object (Strings) and we would like them to be integers so we also changed the type of variables of each column.

In [4]:
#Region and Residence Type to integers:
dictionary_region={"E12000001":1,"E12000002":2,"E12000003":3, "E12000004":4,"E12000005":5,"E12000006":6,"E12000007":7,"E12000008":8,"E12000009":9,"W92000004":10}
dictionary_residence_type={"C":1,"H":2}
#we use the two previous dictionaries created to replace the keys for the values in each column
df_small=df_small.replace({"Region": dictionary_region,"Residence Type":dictionary_residence_type})

#We show the types of data of the columns which is mostly objects
df_small.dtypes




Person ID                    object
Region                        int64
Residence Type                int64
Family Composition           object
Population Base              object
Sex                          object
Age                          object
Marital Status               object
Student                      object
Country of Birth             object
Health                       object
Ethnic Group                 object
Religion                     object
Economic Activity            object
Occupation                   object
Industry                     object
Hours worked per week        object
Approximated Social Grade    object
dtype: object

In [5]:
#we create a list with the column names
column_names=list(df_small.columns)
#create a dictionary
dictionary = dict()
#this for will set the column names as the key of the dictionary and the int class as the value
for item in column_names:
    key = item
    value=int
    
    dictionary[key] = value
#we change the data type of each column using the dictionary created
df_Small = df_small.astype(dictionary)

#show the new data types for the columns
df_Small.dtypes


Person ID                    int32
Region                       int32
Residence Type               int32
Family Composition           int32
Population Base              int32
Sex                          int32
Age                          int32
Marital Status               int32
Student                      int32
Country of Birth             int32
Health                       int32
Ethnic Group                 int32
Religion                     int32
Economic Activity            int32
Occupation                   int32
Industry                     int32
Hours worked per week        int32
Approximated Social Grade    int32
dtype: object

### Data clenaing <a id='data_cleaning'></a>

Given the scope of our project we don't need data for people under 16 neither students living away during term-time. Therefore, we delete the rows corresponding to these cases.

In [6]:
#We replace the -9 values that refer to no code required to the 0 as it will make it easier to plot the graphs
df_Small=df_Small.replace(-9,0)
#delete the rows in which age is under 16, which corresponds to value 1
df_Small = df_Small[~(df_Small["Age"] == 1)]
#delete the rows in wich population base is students living away during term-time, which corresponds to value 2
df_Small = df_Small[~(df_Small["Population Base"] == 2)]


#we define the list variables with all the column names of the dataframe
variables=column_names[1:]
#we reset the index of the dataframe after all these changes
df_Small.reset_index(drop=True,inplace=True)
df_Small



Person ID  Region  Residence Type  Family Composition  \
0         7394816       1               2                   2   
1         7394745       1               2                   5   
2         7395066       1               2                   3   
3         7395329       1               2                   3   
4         7394712       1               2                   3   
...           ...     ...             ...                 ...   
457118    7944868      10               2                   2   
457119    7946020      10               2                   1   
457120    7944310      10               2                   3   
457121    7944768      10               2                   1   
457122    7944959      10               2                   2   

        Population Base  Sex  Age  Marital Status  Student  Country of Birth  \
0                     1    2    6               2        2                 1   
1                     1    1    4               1        2                 1   
2                     1    2    4               1        2                 1   
3                     1    2    2               1        2                 1   
4                     1    1    5               4        2                 1   
...                 ...  ...  ...             ...      ...               ...   
457118                1    1    2               1        2                 1   
457119                1    1    5               1        2                 1   
457120                1    1    3               1        2                 1   
457121                1    2    8               5        2                 1   
457122                1    2    2               2        2                 1   

        Health  Ethnic Group  Religion  Economic Activity  Occupation  \
0            2             1         2                  5           8   
1            1             1         2                  1           8   
2            1             1         1                  1           6   
3            2             1         2                  1           7   
4            1             1         2                  1           1   
...        ...           ...       ...                ...         ...   
457118       1             1         2                  1           4   
457119       4             1         9                  1           8   
457120       2             1         1                  1           7   
457121       3             1         9                  5           9   
457122       2             1         1                  1           7   

        Industry  Hours worked per week  Approximated Social Grade  
0              2                      0                          4  
1              6                      4                          3  
2             11                      3                          4  
3              7                      3                          2  
4              4                      3                          2  
...          ...                    ...                        ...  
457118         9                      3                          2  
457119         8                      3                          3  
457120         4                      3                          4  
457121         2                      0                          4  
457122         4                      1                          4  

[457123 rows x 18 columns]

### Encoding labels

Given that the data is integer encoded we will need dictionaries that associate the label to the value to set the labels in the plots or other implementations in the different functions. 

In [7]:
#Encoding labels
#Dictionary for every variable label and its corresponding value
dict_family_composition={'Resident of a comunal establishment or short-term resident':0,'Not in a family':1, 'Married':2, 'Cohabiting couple':3, 'Lone parent family (male)':4, 'Lone parent family (female)':5, 'Others':6}
dict_residence_type={'Resident in a communal establishment':1,'Not resident in a communal establishment':2}
dict_population_base={'Usual resident':1, 'Short-term resident':3}
dict_sex={'Male':1, 'Female':2}
dict_age={'16 to 24':2, '25 to 34':3, '35 to 44':4, '45 to 54':5, '55 to 64':6, '65 to 74':7, '75 and over':8}
dict_marital_status={'Single':1, 'Married':2, 'Separated':3, 'Divorced':4, 'Widowed':5}
dict_student={'Yes':1, 'No':2}
dict_country_of_birth={'UK':1, 'Non UK':2}
dict_health={'Very good health':1, 'Good health':2, 'Fair health':3, 'Bad health':4, 'Very bad health':5}
dict_ethnic_group={'White':1, 'Mixed':2, 'Asian and Asian British':3, 'Black or Black British':4, 'Chinese or Other ethnic group':5}
dict_religion={'No religion':1, 'Christian':2, 'Buddhist':3, 'Hindu':4, 'Jewish':5, 'Muslim':6, 'Sikh':7, 'Other religion':8, 'Not stated':9}
dict_economic_activity={'Employee':1, 'Self-employed':2, 'Unemployed':3, 'Full-time student':4, 'Retired':5, 'Student':6, 'Looking after home or family':7, 'Long-term sick or disabled':8, 'Others':9}
dict_occupation={'People that never worked':0,'Managers, Directors and Senior Officials':1, 'Professional Occupations':2, 'Associate Professional and Technical Occupations':3, 'Administrative and Secretarial Occupations':4, 'Skilled Trades Occupations':5, 'Caring, Leisure and Other Service Occupations':6, 'Sales and Customer Service Occupations':7, 'Process, Plant and Machine Operatives':8, 'Elementary Occupations':9}
dict_industry={'People never worked':0,'Agriculture, forestry and fishing':1, 'Mining and quarrying; Manufacturing; Electricity, gas, steam and air conditioning system; Water supply':2, 'Construction':3, 'Wholesale and retail trade; Repair of motor vehicles and motorcycles':4, 'Accommodation and food service activities':5, 'Transport and storage; Information and communication':6, 'Financial and insurance activities; Intermediation':7, 'Real estate activities; Professional, scientific and technical activities; Administrative and support service activities':8, 'Public administration and defence; compulsory social security':9, 'Education':10, 'Human health and social work activities':11, 'Other community, social and personal service activities; Private households employing domestic staff; Extra-territorial organisations and bodies':12}
dict_hours_worked_per_week={ 'Not working':0,'Part-time: 15 or less hours worked':1, 'Part-time: 16 to 30 hours worked':2, 'Full-time: 31 to 48 hours worked':3, 'Full-time: 49 or more hours worked':4}
dict_approximated_social_grade={'Resident in a comunal establishment':0,'AB':1, 'C1':2, 'C2':3, 'DE':4}
dict_region={"North East":1,"North West":2,"Yorkshire and the Humber":3,"East Midlands":4,"West Midlands":5,"East of England":6,"London":7,"South East":8,"South West":9,"Wales":10}
#list with all the dictionaries for each variable
dict_variables=[dict_region,dict_residence_type,dict_family_composition, dict_population_base,dict_sex,dict_age,dict_marital_status,dict_student, dict_country_of_birth,dict_health,dict_ethnic_group,dict_religion, dict_economic_activity,dict_occupation,dict_industry,dict_hours_worked_per_week,dict_approximated_social_grade]




### Data visualization and descrpitive analysis <a id='descriptive_analysis'></a>

<p style='text-align: justify;'>Finally, we have defined four different functions that will be used to plot the main descriptive plots of each variable interactively.
    
First, the **generate_pie_chart**, will take the name of a variable and return the pie chart showing the distribution of this variable.</p>




In [8]:
def generate_pie_chart(Variable):
    
        #we plot the pie chart using the value_count which computes the frequency of each value. We also use sort_index 
        #to sort them in order that it will be easier to match the labels in the legend
        plt.pie(df_Small[Variable].value_counts().sort_index(), labels=None,autopct='%1.1f%%')
        #the title of the plot in bold
        plt.title(Variable,weight='bold', size=14)
        #the legend of the plot using the function match_varaible_lables to have the labels of the variable
        plt.legend( match_variable_labels(Variable),loc='center left', bbox_to_anchor=(1.5, 0.5));
        return plt

<p style='text-align: justify;'>Then, the <b>generate_relation_plot</b>, creates a barplot that relates the variable passed as argument and the Hours worked per week variable.</p>

In [9]:
def generate_relation_plot(Variable):
        #a stores a crosstable of two factors which are the variable passed as an argument and Hours worked per week
        a=pd.crosstab(df_Small["Hours worked per week"],df_Small[Variable])
        #we plot the crosstable
        a.plot.bar()
        #plot the legend using the same technique as in generate_pie function 
        plt.legend(match_variable_labels(Variable).keys(),loc='center left', bbox_to_anchor=(1.5, 0.5))
        #we configure the values of the x axis to see the different labels of the variable Hours worked per week
        plt.xticks(list(dict_hours_worked_per_week.values()), list(dict_hours_worked_per_week.keys()),rotation="vertical")
        return plt

<p style='text-align: justify;'>
    The <b>match_variable_labels</b>, is an auxiliar function that will be useful in different parts of the project as it matches the variable with the labels corresponding to the values of the variable. Specifically, it return the list of the labels .</p>

In [10]:
def match_variable_labels(Variable):
    #we implement a for loop that will have a boolean condition to see whether the variable passed is equal to the variable of 
    #the variable list passed in the for loop. If that's the case, we will have the position of the dictionary of the current variable in 
    #the list of dictionaries. This dictionary will be returned as label. In the other function we will use the method .keys() 
    #to get the keys of the dictionaries as lables.
    for i in range(len(variables)):
        if Variable == variables[i]:
            label=dict_variables[i]
            break
    return label

<p style='text-align: justify;'>
    Finally, the <b>descriptive</b> function groups together the other three functions to plot the pie chart of the variable and the bar plot relating the variable with Hours worked per week.</p>

In [11]:
def descriptive(Variable,Execute):
      
    #if the execute variable is true it will execute the function(enter the if)        
    if Execute==True:
        #print the title        
        print("----------------------------------------------------------------\n"+"                 Pie Chart          \n"+"----------------------------------------------------------------\n")
        #show the plot created in the function generate_pie_chart
        generate_pie_chart(Variable).show()
        #print title
        print("----------------------------------------------------------------\n"+"                 Relation with Hours worked per week          \n"+"----------------------------------------------------------------\n")

        #show the plot created in the function generate_relation_plot
        generate_relation_plot(Variable).show()
    else: 
        #if the execute variable is not pressed this message will be displayed
        print("In order to display the plots you need to press the execute button")

In [12]:
#in this cell we use the interactive function from ipywidgets to make the descriptive functions interactive. We can choose 
#whether to execute or not the function by clicking the number and the variable to be plotted from all the variables
w=interactive(descriptive,Variable=variables,Execute=False)
display(w)

interactive(children=(Dropdown(description='Variable', options=('Region', 'Residence Type', 'Family Compositio…


* As we can see from the pie charts, people in the sample are almost equally distributed to what regards regions of origin, gender, age, industry sector, occupation, and approximated social grade.

* London, South East and North West are the regions with most of the full-time job workers: this is probably because of the high cost of life in London and South East, while a lot of steel industries, mines and factories were built in the North-West, and that's why plenty of workers went living there.

* Quite logically, given that feeding and supporting a family requires more money and, most of the time, more hours worked, married people, which corresponds to more than half of the sample, are mostly involved in full time jobs.

* To what regards gender differences, we can see that males and females are equally distributed, but at the same time males are more involved in full time jobs while females are more involved with part-time jobs: this could be related to the fact that women are generally seen as responsible for houseworks and other chores not necessarily related with their jobs.

* People in the sample are also equally split along all the age ranges present in the analysis, with people from 25 to 54 more involved with full-time jobs and people under 25 involved with part-time jobs, given that most people under 25 are probably student, and therefore they don't have time for a full-time job.

* 87.5% of the sample are white people, with the biggest minority being asian people.

* It's interesting to note that "professional occupations" and "associate professional and technical occupations" are the two industry sectors mostly present in full-time jobs in the 31 to 48 hours range, while "Managers, directors and seniors officials" are the majority in 49 or more hours range.

* At the same time, industry types with more hours worked are "Real Estate activities; Professional, Scientific and Technical activities; Administrative and support service activities", "Wholesale and Retail Trade; Repair of motor vehicles and motorcycles", and "Transport and Storage; Information and Communication".

* As a final remark, AB approximated social grade, which corresponds to "higher and intermediate managerial, administrative and professional occupations", is the one with most hours worked; C1 grade follows, corresponding to "supervisory, clerical and junior managerial, administrative, professional occupations".

## Predicting hours worked : Neural Network approach <a id='neural_network'></a>

<p style='text-align: justify;'>As a brief introduction, a Neural Network is a trainable tool that is formed by connecting the different
layers containing neural units, which are the processing elements, that enable the Neural
Network to learn and recognise patterns. Data propagates from
the input layer to the output layer by connecting neurons in that single direction. Within
the process, each neuron receives inputs from previous neurons and transfers outputs to
next neurons. Depending on the strength of each connection, a weight is assigned to it,
turning each neuron into a function of weighted inputs that in turn generate an output
that is used as input to the next neuron. Once we have the
output layer arising from forward feeding, the error of this is calculated by comparing
the outputs generated with the true values of the dependent variable. Then, the error
propagates backwards through the hidden layers until it reaches the input layer to readjust
the weights seeking to minimise the error (”learning”) and that is why it is called back
propagation. The aforementioned adjustments are produced by iteration with the objective of converging to an optimum. Therefore we have to set all the parameters mentioned (number of hidden layers and nuerons, loss-function, optimizer, the activation function and the learning rate) and train our model to get the best results.

First of all we will prepare our data to train and evaluate our model. In order to do so we will one-hot encode all the nominal categorical variables to make each of the values of the variables binary which makes it easier for the model to detect paterns as it helps reduce the bias that the different variable scales may produce.</p> 

In [13]:
#generate the list of variables we are going to one-hot encode
dummy_variables=list(column_names[1:])
#We remove the ordinal categorical variables from the list
dummy_variables.remove('Age')
dummy_variables.remove('Health')
#We remove the variable that will be the target
dummy_variables.remove('Hours worked per week')
#We create the dataset that will be used to feed the neural network which will contain the ordinal variables and the one-hot encoded
#nominal variables together with the target
df_nn=df_Small[['Age','Health','Hours worked per week']]

#We implement a for loop that will one-hot encode each of the nominal variables and put them in the new database
for column in dummy_variables:
    #it will store the arrays of 1s and 0s for each row of the column and add them in the new dataset. The prefix argument is the
    #first part of the new column name.
    dummies = pd.get_dummies(df_Small[column],prefix=column)
    #create the new columns (as many as values could take the variable when was integer encoded)in the new dataset
    df_nn[dummies.columns] = dummies


<p style='text-align: justify;'>Now that we have the dataset prepared, we have to separate the data in the features that will be used to predict the hours worked per week and also in the target variable itself. After this we are going to split the features and the target variables between the training set and the test set. Therefore, we are going to have the train set of features (x_train), the train set  of variables(y_train), the test set of features(x_test) and the test set of targets (y_test).</p> 

In [42]:
#We set the list of features excluding the target variable
features=list(df_nn.columns)
features.remove('Hours worked per week')

#We set the features data to x and the target variable to y
x=df_nn[features]
y=df_nn[["Hours worked per week"]]

#We randomly split the data putting 20% of the observations into the test set and 80% in the training set 
x_train, x_test, y_train, y_test= train_test_split(x,y,random_state=50,test_size=0.20)

#After claning and transforming the data we think there is no need to standarize our cateogrical variables. Moreover, 
#we have tried standarizing and there were no relevant changes in prediction.




<p style='text-align: justify;'>Once the data is prepared we have to configure the neural network model. To do so we are goig to use the keras package from tensorflow. Concretely, we have decided to nest three different for loops that will produce 8 different neural networks. This is a way of doing a little bit of parameter tunning. We could have choosen more options in each for and implement more for loops in order to improve our tuning but it will have taken too much time. Nonetheless, we have followed a trial and error method before doing this parameter tuning in order to choose the other parameters. Each of these neural networks will be stored in an array. Concretely, we will store the accruacy metric for the validation set of each neural network, the list of the validation loss, the trainning loss, the validation accuracy and the trainning accuracy of each iteration when training the different neural networks to be able to plot the curves and analyse if there is over or underfitting. <b>BE AWARE THAT NEXT CELL IS GOING TO TRAIN 8 NEURAl NETWORKS WHICH CAN TAKE UP TO 50 MINUTES</b>.</p>

In [15]:
#we will set to options of the activation functions which will be the relu and the sigmoid functions
activation_function = ['relu', 'sigmoid']
#the first hidden layer will have 100 or 80 neurons
neurons_1 = [100, 80]
#the second hidden layer will have 50 or 40 neurons
neurons_2 = [50,40]
#we initialize the array that will store the data of each neural network
results = []

#We will have 8 neural networks given that we have two options per loop (2^3)
#Activations will take values 'relu' and generate 4 neural networks and then 'sigmoid'and generate 4 more neural networks
for activations in activation_function:
    #neuron_1 will take 100 and 80 values
    for neuron_1 in neurons_1:
        #neuron_2 will take 50 and 40 values
        for neuron_2 in neurons_2:
            # Train the model
            #we set the optimizer as 'Adam',which is a stochastic gradient descent method that is based on adaptive 
            #estimation of first-order and second-order moments. We also set the learning rate 
            opt = tf.keras.optimizers.Adam(learning_rate=0.00001)
            #initialize the model as Sequential which groups different layers 
            model=models.Sequential()
            #we add the input layer and 1 hidden layer with the neurons (100 or 80) and the actvation function(relu or sigmoid)
            model.add(layers.Dense(neuron_1,activation=activations,input_shape=(x_train.shape[1],)))
            #we add the second hidden layer with the neurons (50 or 40) and the actvation function(relu or sigmoid)
            model.add(layers.Dense(neuron_2, activation=activations))
            #we add the output layer with 5 neurons given that we have 5 cateogires with the softmax function which gives a probabilitie
            #for every category or neuron
            model.add(layers.Dense(5,activation='softmax'))
            #we compile the model and set the loss function to 'sparse_categorical_corssentropy', that suits our case as our target variable
            #is integer encoded. We set the optimizer as explained before and we choose the accuracy as the metric to evaluate the performance.
            model.compile(loss='sparse_categorical_crossentropy', optimizer=opt ,metrics=['accuracy'])
            #now we  fit the model with the data and train the model with 100 epochs and 90 as batch size. Epochs are the number of times that 
            #the model will go trough the entire training set in the training process whereas batches are the number of training samples to work through 
            #before updating the model. Finally, we set to 20% the proportion of data from the training set that will be part of the validation set used to
            #control for over or under fitting and for the parameter tuning. Moreover, we store the whole trainning process in the varaible history
            history=model.fit(x_train,y_train,epochs=100,batch_size=90,validation_split=0.20)
            #we also store the predictions that the model does with the features of the training set
            predictions=model.predict(x_train)
            y_predictions=[]
            #this for will select the maxium from the array of 5 probabilities for each category that has been predicted and append it to the list of predictions
            for i in range(predictions.shape[0]):
                maximum = np.argmax(predictions[i])
                y_predictions.append(maximum)
            
            
            
            # Append the current results trough a dictionary
            results.append({
                #we round the accuracy and multiply it by 100
                'Validation accuracy': round(history.history['val_accuracy'][-1]*100, 2),
                'activation_function': activations,
                'neuron_1': neuron_1,
                'neuron_2': neuron_2,
                'loss':history.history['loss'],
                'val_loss':history.history['val_loss'],
                'accuracy':history.history['accuracy'],
                'val_accuracy':history.history['val_accuracy'],
                'predictions':y_predictions
            })
            
            
            




Epoch 1/100
3251/3251 [==============================] - 5s 1ms/step - loss: 1.2826 - accuracy: 0.5332 - val_loss: 1.0478 - val_accuracy: 0.6830
Epoch 2/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.8845 - accuracy: 0.7162 - val_loss: 0.7497 - val_accuracy: 0.7450
Epoch 3/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.6802 - accuracy: 0.7568 - val_loss: 0.6270 - val_accuracy: 0.7668
Epoch 4/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.6034 - accuracy: 0.7682 - val_loss: 0.5828 - val_accuracy: 0.7715
Epoch 5/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5752 - accuracy: 0.7713 - val_loss: 0.5662 - val_accuracy: 0.7731
Epoch 6/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5639 - accuracy: 0.7722 - val_loss: 0.5587 - val_accuracy: 0.7741
Epoch 7/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5583 - accuracy: 0.7727 - val_loss: 0.5546 - val_ac

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5285 - accuracy: 0.7821 - val_loss: 0.5284 - val_accuracy: 0.7837
Epoch 58/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5282 - accuracy: 0.7821 - val_loss: 0.5281 - val_accuracy: 0.7836
Epoch 59/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5279 - accuracy: 0.7822 - val_loss: 0.5279 - val_accuracy: 0.7837
Epoch 60/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5277 - accuracy: 0.7824 - val_loss: 0.5277 - val_accuracy: 0.7841
Epoch 61/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5274 - accuracy: 0.7823 - val_loss: 0.5275 - val_accuracy: 0.7839
Epoch 62/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5271 - accuracy: 0.7824 - val_loss: 0.5273 - val_accuracy: 0.7840
Epoch 63/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5269 - accuracy: 0.7825 - val_loss: 0.5272 - val_accuracy

Epoch 13/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5508 - accuracy: 0.7745 - val_loss: 0.5492 - val_accuracy: 0.7759
Epoch 14/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5500 - accuracy: 0.7748 - val_loss: 0.5485 - val_accuracy: 0.7756
Epoch 15/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5493 - accuracy: 0.7752 - val_loss: 0.5479 - val_accuracy: 0.7764
Epoch 16/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5487 - accuracy: 0.7754 - val_loss: 0.5472 - val_accuracy: 0.7767
Epoch 17/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5480 - accuracy: 0.7757 - val_loss: 0.5468 - val_accuracy: 0.7775
Epoch 18/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5474 - accuracy: 0.7760 - val_loss: 0.5460 - val_accuracy: 0.7773
Epoch 19/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5468 - accuracy: 0.7761 - val_loss: 0.5455 -

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5275 - accuracy: 0.7824 - val_loss: 0.5278 - val_accuracy: 0.7839
Epoch 70/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5273 - accuracy: 0.7824 - val_loss: 0.5274 - val_accuracy: 0.7842
Epoch 71/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5270 - accuracy: 0.7825 - val_loss: 0.5273 - val_accuracy: 0.7841
Epoch 72/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5268 - accuracy: 0.7828 - val_loss: 0.5270 - val_accuracy: 0.7841
Epoch 73/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5265 - accuracy: 0.7826 - val_loss: 0.5268 - val_accuracy: 0.7843
Epoch 74/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5263 - accuracy: 0.7828 - val_loss: 0.5265 - val_accuracy: 0.7841
Epoch 75/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5260 - accuracy: 0.7829 - val_loss: 0.5265 - val_accuracy

Epoch 25/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5422 - accuracy: 0.7781 - val_loss: 0.5407 - val_accuracy: 0.7796
Epoch 26/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5416 - accuracy: 0.7782 - val_loss: 0.5400 - val_accuracy: 0.7801
Epoch 27/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5411 - accuracy: 0.7784 - val_loss: 0.5396 - val_accuracy: 0.7800
Epoch 28/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5406 - accuracy: 0.7785 - val_loss: 0.5391 - val_accuracy: 0.7808
Epoch 29/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5402 - accuracy: 0.7786 - val_loss: 0.5387 - val_accuracy: 0.7804
Epoch 30/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5397 - accuracy: 0.7789 - val_loss: 0.5383 - val_accuracy: 0.7804
Epoch 31/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5392 - accuracy: 0.7790 - val_loss: 0.5378 -

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5247 - accuracy: 0.7837 - val_loss: 0.5251 - val_accuracy: 0.7846
Epoch 82/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5245 - accuracy: 0.7839 - val_loss: 0.5249 - val_accuracy: 0.7845
Epoch 83/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5243 - accuracy: 0.7837 - val_loss: 0.5248 - val_accuracy: 0.7846
Epoch 84/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5241 - accuracy: 0.7840 - val_loss: 0.5246 - val_accuracy: 0.7843
Epoch 85/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5240 - accuracy: 0.7839 - val_loss: 0.5244 - val_accuracy: 0.7845
Epoch 86/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5238 - accuracy: 0.7839 - val_loss: 0.5243 - val_accuracy: 0.7848
Epoch 87/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5236 - accuracy: 0.7838 - val_loss: 0.5242 - val_accuracy

Epoch 37/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5402 - accuracy: 0.7790 - val_loss: 0.5389 - val_accuracy: 0.7805
Epoch 38/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5398 - accuracy: 0.7793 - val_loss: 0.5385 - val_accuracy: 0.7810
Epoch 39/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5394 - accuracy: 0.7793 - val_loss: 0.5385 - val_accuracy: 0.7806
Epoch 40/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5391 - accuracy: 0.7796 - val_loss: 0.5379 - val_accuracy: 0.7811
Epoch 41/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5387 - accuracy: 0.7798 - val_loss: 0.5375 - val_accuracy: 0.7812
Epoch 42/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5384 - accuracy: 0.7798 - val_loss: 0.5372 - val_accuracy: 0.7815
Epoch 43/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5380 - accuracy: 0.7800 - val_loss: 0.5368 -

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5266 - accuracy: 0.7831 - val_loss: 0.5262 - val_accuracy: 0.7843
Epoch 94/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5264 - accuracy: 0.7831 - val_loss: 0.5260 - val_accuracy: 0.7846
Epoch 95/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5263 - accuracy: 0.7832 - val_loss: 0.5259 - val_accuracy: 0.7844
Epoch 96/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5261 - accuracy: 0.7833 - val_loss: 0.5257 - val_accuracy: 0.7843
Epoch 97/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5259 - accuracy: 0.7833 - val_loss: 0.5256 - val_accuracy: 0.7847
Epoch 98/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5258 - accuracy: 0.7835 - val_loss: 0.5254 - val_accuracy: 0.7841
Epoch 99/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5257 - accuracy: 0.7834 - val_loss: 0.5253 - val_accuracy

Epoch 49/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5572 - accuracy: 0.7726 - val_loss: 0.5559 - val_accuracy: 0.7732
Epoch 50/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5571 - accuracy: 0.7728 - val_loss: 0.5558 - val_accuracy: 0.7732
Epoch 51/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5570 - accuracy: 0.7727 - val_loss: 0.5556 - val_accuracy: 0.7733
Epoch 52/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5568 - accuracy: 0.7727 - val_loss: 0.5555 - val_accuracy: 0.7733
Epoch 53/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5567 - accuracy: 0.7729 - val_loss: 0.5554 - val_accuracy: 0.7734
Epoch 54/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5565 - accuracy: 0.7728 - val_loss: 0.5553 - val_accuracy: 0.7736
Epoch 55/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5564 - accuracy: 0.7729 - val_loss: 0.5551 -

Epoch 5/100
3251/3251 [==============================] - 5s 2ms/step - loss: 0.9667 - accuracy: 0.7107 - val_loss: 0.9129 - val_accuracy: 0.7187
Epoch 6/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.8705 - accuracy: 0.7245 - val_loss: 0.8273 - val_accuracy: 0.7325
Epoch 7/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.7941 - accuracy: 0.7355 - val_loss: 0.7596 - val_accuracy: 0.7419
Epoch 8/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.7342 - accuracy: 0.7426 - val_loss: 0.7068 - val_accuracy: 0.7471
Epoch 9/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.6880 - accuracy: 0.7474 - val_loss: 0.6668 - val_accuracy: 0.7512
Epoch 10/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.6535 - accuracy: 0.7516 - val_loss: 0.6375 - val_accuracy: 0.7551
Epoch 11/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.6287 - accuracy: 0.7549 - val_loss: 0.6169 - val_

3251/3251 [==============================] - 5s 1ms/step - loss: 0.5560 - accuracy: 0.7733 - val_loss: 0.5546 - val_accuracy: 0.7741
Epoch 62/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5559 - accuracy: 0.7733 - val_loss: 0.5545 - val_accuracy: 0.7741
Epoch 63/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5558 - accuracy: 0.7734 - val_loss: 0.5544 - val_accuracy: 0.7742
Epoch 64/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5557 - accuracy: 0.7734 - val_loss: 0.5545 - val_accuracy: 0.7748
Epoch 65/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5556 - accuracy: 0.7734 - val_loss: 0.5543 - val_accuracy: 0.7741
Epoch 66/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5556 - accuracy: 0.7734 - val_loss: 0.5542 - val_accuracy: 0.7744
Epoch 67/100
3251/3251 [==============================] - 5s 1ms/step - loss: 0.5555 - accuracy: 0.7734 - val_loss: 0.5541 - val_accuracy

Epoch 17/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5869 - accuracy: 0.7679 - val_loss: 0.5827 - val_accuracy: 0.7690
Epoch 18/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5817 - accuracy: 0.7689 - val_loss: 0.5781 - val_accuracy: 0.7698
Epoch 19/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5776 - accuracy: 0.7691 - val_loss: 0.5745 - val_accuracy: 0.7709
Epoch 20/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5744 - accuracy: 0.7695 - val_loss: 0.5717 - val_accuracy: 0.7713
Epoch 21/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5718 - accuracy: 0.7699 - val_loss: 0.5694 - val_accuracy: 0.7716
Epoch 22/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5697 - accuracy: 0.7702 - val_loss: 0.5675 - val_accuracy: 0.7717
Epoch 23/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5680 - accuracy: 0.7702 - val_loss: 0.5660 -

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5548 - accuracy: 0.7736 - val_loss: 0.5534 - val_accuracy: 0.7743
Epoch 74/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5547 - accuracy: 0.7736 - val_loss: 0.5533 - val_accuracy: 0.7745
Epoch 75/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5546 - accuracy: 0.7735 - val_loss: 0.5533 - val_accuracy: 0.7748
Epoch 76/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5545 - accuracy: 0.7737 - val_loss: 0.5531 - val_accuracy: 0.7745
Epoch 77/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5544 - accuracy: 0.7737 - val_loss: 0.5531 - val_accuracy: 0.7741
Epoch 78/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5543 - accuracy: 0.7736 - val_loss: 0.5530 - val_accuracy: 0.7746
Epoch 79/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5543 - accuracy: 0.7737 - val_loss: 0.5529 - val_accuracy

Epoch 29/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5622 - accuracy: 0.7713 - val_loss: 0.5606 - val_accuracy: 0.7721
Epoch 30/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5617 - accuracy: 0.7714 - val_loss: 0.5602 - val_accuracy: 0.7720
Epoch 31/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5613 - accuracy: 0.7715 - val_loss: 0.5597 - val_accuracy: 0.7718
Epoch 32/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5609 - accuracy: 0.7716 - val_loss: 0.5593 - val_accuracy: 0.7718
Epoch 33/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5605 - accuracy: 0.7716 - val_loss: 0.5590 - val_accuracy: 0.7723
Epoch 34/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5602 - accuracy: 0.7717 - val_loss: 0.5586 - val_accuracy: 0.7723
Epoch 35/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5599 - accuracy: 0.7717 - val_loss: 0.5584 -

3251/3251 [==============================] - 4s 1ms/step - loss: 0.5538 - accuracy: 0.7739 - val_loss: 0.5523 - val_accuracy: 0.7745
Epoch 86/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5537 - accuracy: 0.7738 - val_loss: 0.5523 - val_accuracy: 0.7749
Epoch 87/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5536 - accuracy: 0.7739 - val_loss: 0.5522 - val_accuracy: 0.7749
Epoch 88/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5535 - accuracy: 0.7739 - val_loss: 0.5521 - val_accuracy: 0.7748
Epoch 89/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5535 - accuracy: 0.7739 - val_loss: 0.5521 - val_accuracy: 0.7750
Epoch 90/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5534 - accuracy: 0.7740 - val_loss: 0.5520 - val_accuracy: 0.7750
Epoch 91/100
3251/3251 [==============================] - 4s 1ms/step - loss: 0.5533 - accuracy: 0.7740 - val_loss: 0.5519 - val_accuracy

Now that we have the data sorted for each neural network, we can do a table to compare them based on the accuracy of the validation set that is what has to be used in the hyperparameter tunning.

In [27]:
# Convert the array to Pandas DataFrame 
comparison = pd.DataFrame(results)
#add the model column
comparison['Model']=['Model 1','Model 2', 'Model 3','Model 4','Model 5', 'Model 6','Model 7','Model 8']
#keep only the important variables
result1=comparison[['Model','Validation accuracy','activation_function','neuron_1','neuron_2']]
result1

Model  Validation accuracy activation_function  neuron_1  neuron_2
0  Model 1                78.47                relu       100        50
1  Model 2                78.48                relu       100        40
2  Model 3                78.50                relu        80        50
3  Model 4                78.46                relu        80        40
4  Model 5                77.53             sigmoid       100        50
5  Model 6                77.53             sigmoid       100        40
6  Model 7                77.52             sigmoid        80        50
7  Model 8                77.51             sigmoid        80        40

We can observe the characteristics of each model but we have to analyse each of them more in depth to see if there is over or underfitting.

In [28]:
#we generate the dictionary to know the index of each model
dict_models={'Model 1':0,'Model 2':1, 'Model 3':2,'Model 4':3,'Model 5':4, 'Model 6':5,'Model 7':6,'Model 8':7}
#generate a list with the name of each model
model_num=list(dict_models.keys())

<p style='text-align: justify;'>We are going to create three different functions to plot the results of each model iteractively.

First, the **overfitting_plots function**, creates a figure with two subbplots. In the first sublot we can see the curves of the training loss and the validation-loss. This function takes the index of the model as argument and returns the figure.</p>

In [29]:
def overfitting_plots(option):
    #we create two subplots of a certain dimension each    
    fig, (ax1, ax2) = plt.subplots(1,2, figsize=(20,8))
    #we define the variable loss as the array generated by the loss function when iterating in the training process extracted form the comparison data frame.
    loss =comparison.iloc[option]['loss']
    #the same as before but with the validation set
    val_loss = comparison.iloc[option]['val_loss']
    #we set de epochs as the length of the list
    epochs = range(1, len(loss) + 1)
    #we plot the epochs in the x axis and the loss in the y axis in yellow
    ax1.plot(epochs, loss, 'y', label='Training loss')
    #same as before with the validation loss in red
    ax1.plot(epochs, val_loss, 'r', label='Validation loss')
    #put the title of the subplot
    ax1.title.set_text('Training and validation loss')
    #set the x and the y label
    ax1.set_xlabel('Epochs')
    ax1.set_ylabel('Loss')
    #plot the legend
    ax1.legend()
    #we do the second plot as the first but changing the loss for the accuracy.
    accuracy =comparison.iloc[option]['accuracy']
    val_accuracy =comparison.iloc[option]['val_accuracy']
    epochs = range(1, len(loss) + 1)
    ax2.plot(epochs, accuracy, 'y', label='Accuracy')
    ax2.plot(epochs, val_accuracy, 'r', label='Validation Accuracy')
    ax2.title.set_text('Training and validation Accuracy')
    ax2.set_xlabel('Epochs')
    ax2.set_ylabel('Accuracy')
    ax2.legend()
    return plt

<p style='text-align: justify;'>
    Second, the <b>observed_against_predicted</b>  function, takes the index of the model in the comparison table as argument and returns a barplot with the observed value against the predicted value for the first 20 observations of the training set.</p>

In [43]:
#dataframe with the Observed column
observed_predicted=y_train.copy()
def observed_against_predicted(option):
    #we first construct a dataframe with the column of the observed hours worked per week and the predicted ones(two columns)    
    observed_predicted['Predicted']=comparison.iloc[option]['predictions']
    observed_predicted.rename(columns = {'Hours worked per week':'Observed'},inplace = True)
    #we reset the index to go from 0 onwards 
    observed_predicted.reset_index(drop=True,inplace=True)
    #we sum one to all the values of the dataframe to be able to plot the vaalue 0
    observed_predicted2=observed_predicted+1
    #we plot only the 20 first values of the dataframe
    table=observed_predicted2.head(20)
    table.plot.bar()
    #we switch the values of the y axis to make them the cateogries of the hours worked per week variable
    plt.yticks([0,1,2,3,4,5],["","Not working","Part-time: 15 or less hours worked","Part-time: 16 to 30 hours worked","Full-time: 31 to 48 hours worked","Full-time: 49 or more hours worked"])
    plt.show()

<p style='text-align: justify;'> Last, the <b>plot_model_results</b>, takes the name of the model as argument and has the previous two functions nested so that it plots both the over or underfitting  plots and the comaprison of predicted against observed values of the model choosed. Moreover, it prints a string that indicates the accuracy and other characteristics of the model.</p>

In [44]:
def plot_model_results(model):
    #using the previous dictionary  we find the index with the name of the model
    option=dict_models[model]
    #print the title
    print('\n')
    print('\033[1m'+"                                                           " +model+ ' Results:'+'\033[0m')
    #plot the curves of over and underfitting
    overfitting_plots(option).show()
    #prints a title to separate both plots
    print('--------------------------------------------------------')
    print('  Hours worked per week: Observed against Predicted  ')
    print('--------------------------------------------------------')
    #plots the arplot of the values predicted and observed
    observed_against_predicted(option)
    #prints the string with the data from the model
    print('For this model we have used '+ str(comparison.iloc[option]['activation_function']) +' as activation function and the first hidden layer had '+str(comparison.iloc[option]['neuron_1']) 
          + ' neurons while the second hidden layer had '+str(comparison.iloc[option]['neuron_2']) + ' neurons. The accuracy of this model measured with the validation set has been: '+ str(comparison.iloc[option]['Validation accuracy']) + '%.')

In [45]:
#we finally make the last function described interactively and we set all the model names as the options
#to choose in the parameter of the function
res=interactive(plot_model_results,model=model_num)
display(res)    

interactive(children=(Dropdown(description='model', options=('Model 1', 'Model 2', 'Model 3', 'Model 4', 'Mode…

In [22]:
#in the next variable we store which is the model with the best accuracy in the validation set
chosen=comparison.loc[comparison['Validation accuracy'].idxmax()]["Model"]
#we also store the parameters of the best model for the bootstraping section
chosen_neuron_1=comparison.loc[comparison['Validation accuracy'].idxmax()]["neuron_1"]
chosen_neuron_2=comparison.loc[comparison['Validation accuracy'].idxmax()]["neuron_2"]
chosen_activation_function=comparison.loc[comparison['Validation accuracy'].idxmax()]["activation_function"]
print("Given that none of the previous models has overfitting, we are going to choose the best model based on the validation set accuracy. In this case this is "+chosen+".")

Given that none of the previous models has overfitting, we are going to choose the best model based on the validation set accuracy. In this case this is Model 3.


The model chosen will be compared to the best model of the second approach, which will be carried in the next section,  by the accuracy of each model when using out of sample data.

## Logistic Multinomial Regression  <a id='logistic_regression'></a>

<p style='text-align: justify;'>Multinomial logistic regression is a simple extension of binary logistic regression that allows for more than two categories of the dependent or outcome variable. Like binary logistic regression, multinomial logistic regression uses maximum likelihood estimation to evaluate the probability of categorical membership.

The LogisticRegression class can be used for a multinomial logistic regression if we specify the argument multi_class as multinomial. In order to be able to model multinomial target the solver of the model has to be set as lbfgs as well.

The multinomial logistic regression model will be fit using cross-entropy loss and will predict the integer value for each integer encoded class label. This loss function is the same as the one used in the neural network model.</p>



In [23]:
#we first initilize the model with the previosuly mentioned parameters to adapt the binary logistic regression to a multiple logistic regression.
model = LogisticRegression(multi_class='multinomial', solver='lbfgs')

    
    

In [47]:
y_train

Hours worked per week
436001                      3
105423                      3
18737                       0
319520                      3
110075                      2
...                       ...
441633                      3
153709                      2
239499                      0
103904                      0
407216                      3

[365698 rows x 1 columns]

In [48]:
#we now train the model
model.fit(x_train, y_train)

KeyboardInterrupt: 

In [ ]:
#We make predictions using the test set 
logistic_pred=model.predict(x_test)
#we use the previously predictions to get the accuracy of the model when using data out of sample
accuracy_score(y_test[['Hours worked per week']],logistic_pred)

Now we are going to repeat the same process but including different levels of penalty by implementing a for loop.Specifically, penalty imposes pressure on the model to seek smaller model weights. There are three types of penalty but we have chosen the L2 penalty that adds the (weighted) sum of the squared coefficients to the loss function. A weighting of the coefficients(C) can be used that reduces the strength of the penalty from full penalty to a very slight penalty. This will allow us to compare them and choose the best one based on the test set accuracy. The choosen one will be the one used to be compared with the neural network model choosen in the [previous section](#neural_network).

In [ ]:
#we initialize the dictionary to store each alternative model
logistic_models = dict()
#we implement a for loop to create a model for each penalty alternative. As clarification, penalty = 1 – C. That is why we add
#the parameter C.
for p in [0.0, 0.0001, 0.001, 0.01, 0.1, 1.0]:
    # create name for model
    key = '%.4f' % p
    #the for loop will enter thos if when the penalty of the model equals 0
    if p == 0.0:
        # no penalty in this case
        logistic_models[key] = LogisticRegression(multi_class='multinomial', solver='lbfgs', penalty='none')
    else:
        #penalty equal to p in this case
        logistic_models[key] = LogisticRegression(multi_class='multinomial', solver='lbfgs', penalty='l2', C=p)

In [ ]:
#we initialize the dictionary in which we will store the test set accuracy for each of the models previously created
accuracies=dict()
#we implement a for loop that will compute and store the test accuracy for each model
for key,value in logistic_models.items():
    #we train each model
    value.fit(x_train,y_train)
    #once the model is trained we make the predictions based on the test set
    predictions=value.predict(x_test)
    #we compute the test accuracy for each model and store it in the dictionary
    accuracies['Model with pealty: '+key]=accuracy_score(y_test[['Hours worked per week']],predictions)
    
    

In [ ]:
#we create the data frame to have  a table with each model and its accuracy by using the ditionary accuracies previously created
logit_comparison=pd.DataFrame({"Model":list(accuracies.keys()),"Accuracies":list(accuracies.values())})
logit_comparison

In [ ]:
print("Given the table shown above we consider the "+logit_comparison.iloc[logit_comparison["Accuracies"].idxmax()]["Model"] +" to be the best.")

## A comparison of approaches: Bootstrapping <a id='bootstrapping'></a>

In [ ]:
#we define a function to work with the same neural network every loop. THis neural network is the chosen in the neural network 
#section. The parameters of this neural network are "chosen_activation_function","chosen_neuron_1",chosen_neuron_2.
def compile_neural_model():
    
    opt = tf.keras.optimizers.Adam(learning_rate=0.00001)
    neural_model=models.Sequential()
    neural_model.add(layers.Dense(chosen_neuron_1,activation=chosen_activation_function,input_shape=(x_train.shape[1],)))
    neural_model.add(layers.Dense(chosen_neuron_2, activation=chosen_activation_function))
    neural_model.add(layers.Dense(5,activation='softmax'))
    neural_model.compile(loss='sparse_categorical_crossentropy', optimizer=opt ,metrics=['accuracy'])
    return neural_model
#we define the next function to work with the same logistic multinomial regression every loop
def compile_logistic_model():
    logistic_model = LogisticRegression(multi_class='multinomial', solver='lbfgs',penalty='none')
    return logistic_model

In [ ]:
#split data change the random_state to make each split different
#initialize the array to store the accuracies for every neural and for every logistic multinomial regression
neural_accuracies=[]
logistic_accuracies=[]
#set 10 different random_state which values are irrelevant as long as they are different
random_states=[10,20,30,40,50,60,70,80,90,100]
#one loop for every random state to make our analysis more robust by training and predicting with different splits which
#gives consistency
for i in random_states:
    #we first split the data that will be used in both models. Each round the split will be different so the train and test set
    #are different every time
    x_train, x_test, y_train, y_test= train_test_split(x,y,random_state=i,test_size=0.20)
    
    #we compile each model as it was never trained
    neural_network=compile_neural_model()
    logistic_model=compile_logistic_model()
    #we train each model
    neural_network.fit(x_train,y_train,epochs=100,batch_size=90,validation_split=0.20)
    logistic_model.fit(x_train, y_train)
    #we evaluate the neural network model
    #we also store the predictions that the model does with the features of the test set
    predictions=neural_network.predict(x_test)
    y_predictions=[]
    #this for will select the maxium from the array of 5 probabilities for each category that has been predicted and append it 
    #to the list of predictions        
    for i in range(predictions.shape[0]):
            maximum = np.argmax(predictions[i])
            y_predictions.append(maximum)
    
    
    
    neural_stat =accuracy_score(y_test[['Hours worked per week']],y_predictions) 
    neural_accuracies.append(neural_stat)
    #we evaluate the logistic model
    logistic_predictions=value.predict(x_test)
    logistic_stat=accuracy_score(y_test[['Hours worked per week']],logistic_predictions)
    logistic_accuracies.append(logistic_stat)

    



<p style='text-align: justify;'>In the next cell we will use each of the accuracies arrays created using the bootstraping method to find the confidence interval of the neural network predictions accuracies and the logistic multinomial regression prediction accuracies. To do so we are going to plot the boxplot of the accuracies distributions as well as the confidence intervals which can be done by finding the 
correspondent percentiles of the accuracies array. We use the numpy percentile method. Firstwe chose the confidence level which we choose to be 95, thus we find the 2.5 and the 97.5 percentiles.</P>

In [ ]:
neural_confidence_interval_95=np.percentile(neural_accuracies,[100*(1-0.95)/2,100*(1-(1-0.95)/2)])
logistic_confidence_interval_95=np.percentile(logistic_accuracies,[100*(1-0.95)/2,100*(1-(1-0.95)/2)])
models_accuracies_comparison=pd.DataFrame({"Neural Network Accuracies":neural_accuracies,"Logistic Regression Accuracies":logistic_accuracies})
sn.boxplot(data=models_accuracies_comparison, orient="h", palette="Set2")
plt.xlabel('accuracy')
plt.title('Accuracies distribution per model with 10 samples')
plt.show()
print("The accuracy of the Neural Network model predictions goes from "+str(round(neural_confidence_interval_95[0]*100,2)) +"% to "+str(round(neural_confidence_interval_95[1]*100,2)) +"% in 95% of the times. Regarding the Logistic Multinomial Regression model, the accuracy of its predictions goes from "+str(round(logistic_confidence_interval_95[0]*100,2))+"% to "+str(round(logistic_confidence_interval_95[1]*100,2)) +"% in 95% of the times.")

Given this results we can say that the **Neural Network model does a slightly better job** in predicting the hours worked per week which can be seen in the difference between the predictions accuracy confidence interval.


# Robustness analysis (0.5 points)



<p style='text-align: justify;'>We consider that the procedure of our analysis is quite robust given that we have performed diferent comparisons at an intra model stage and between approaches. We first performed a comparison between neural networks when doing the hyperparameter tunning by looking at the validation set accuracy among 8 different models. Secondly, we have compared the different Logistic Multinomial Regression models by setting different penalty degrees and looking at the accuracy when using data out of sample. Lastly, we have used a bootstraping method to take samples of the same neural network and the same Logistic Multinomial Regression accuracy when being trained with different data and also tested with different test sets which has allowed us to give a confidence interval of the predictions accuracy for each model. Nonetheless, each of this steps has been performed with few samples which weakness our study robustness. This las point will be extended in the discussion.</p>

# Discussion and conclusion (0.5 points)





<p style='text-align: justify;'>All things considered I thing that our study can give an answer to our research question. The answer to the first part of our research question, which consisted in estimating to what degreee can personal characteristics help predicting the level of hours worked per week in the UK, is that the personal characteristic can help to reach an accuracy in the predictions between 78% and 78.5% in 95% of times based on the best model we could configure.  Regarding the second part of the research question, which was to detemine which method does a better job in predicting the hours worked per week; the Neural Network or the Logistic Multinomial Regression, we have found that the Neural network does a slightly better job as the confidence interval is better.

Nonetheless, we consider our study to have some weaknesses that make our results not robust enough. This particular weaknesses can be divided into two groups. First of all, the time and memory constraint leaded to a not consistent enough hyperparameter tunning that avoid us from finding a better neural network model as with more time and memory we could have increased the alternatives for each parameter. The same happened with the bootstraping section to compare both model as we just did a sample of ten models for alternatives which is too low. The second group of weaknesses correspond to the data that we have got which is only categorical leaving out some numerial features like income, number of people responsible of and others that may increase the predictive performance. 

To conclude, we can say that we have conducted quite a complete analysis that can be further developed by imporving some of the weaknesses previously mentioned.</p>